### PR TITLE
Do not run resource registry stuff if migrations are not finished

### LINKED ISSUE #555: Reverse migrations are not possible because of resource_registry
Reproducer is to run the bootstrap script, then:

```
python manage.py migrate dab_rbac zero
```

that gives:

```
2024-08-06 16:29:10,176 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Including URLS from ansible_base.api_documentation.urls
2024-08-06 16:29:10,176 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Including URLS from ansible_base.api_documentation.urls
2024-08-06 16:29:10,216 DEBUG    []  ansible_base.authentication.urls Loaded URLS from ansible_base.authentication.authenticator_plugins.saml
2024-08-06 16:29:10,216 DEBUG    []  ansible_base.authentication.urls Loaded URLS from ansible_base.authentication.authenticator_plugins.saml
2024-08-06 16:29:10,243 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Including URLS from ansible_base.authentication.urls
2024-08-06 16:29:10,243 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Including URLS from ansible_base.authentication.urls
2024-08-06 16:29:10,243 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Module ansible_base.rest_filters does not specify urls.py
2024-08-06 16:29:10,243 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Module ansible_base.rest_filters does not specify urls.py
2024-08-06 16:29:10,244 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Including URLS from ansible_base.jwt_consumer.urls
2024-08-06 16:29:10,244 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Including URLS from ansible_base.jwt_consumer.urls
2024-08-06 16:29:10,245 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Including URLS from ansible_base.resource_registry.urls
2024-08-06 16:29:10,245 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Including URLS from ansible_base.resource_registry.urls
2024-08-06 16:29:10,245 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Module ansible_base.rest_pagination does not specify urls.py
2024-08-06 16:29:10,245 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Module ansible_base.rest_pagination does not specify urls.py
2024-08-06 16:29:10,247 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Including URLS from ansible_base.rbac.urls
2024-08-06 16:29:10,247 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Including URLS from ansible_base.rbac.urls
2024-08-06 16:29:10,247 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Including URLS from ansible_base.oauth2_provider.urls
2024-08-06 16:29:10,247 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Including URLS from ansible_base.oauth2_provider.urls
2024-08-06 16:29:10,248 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Including URLS from ansible_base.activitystream.urls
2024-08-06 16:29:10,248 DEBUG    []  ansible_base.lib.dynamic_config.dynamic_urls Including URLS from ansible_base.activitystream.urls
Operations to perform:
  Unapply all migrations: dab_rbac
Running migrations:
  Rendering model states... DONE
  Unapplying test_app.0014_autoextrauuidmodel_manualextrauuidmodel_and_more... OK
  Unapplying test_app.0013_alter_manageduser_managers_alter_user_managers... OK
  Unapplying test_app.0012_encryptionjsonmodel... OK
  Unapplying test_app.0011_publicdata... OK
  Unapplying test_app.0010_manageduser... OK
  Unapplying test_app.0009_city_state... OK
  Unapplying test_app.0008_secretcolor... OK
  Unapplying test_app.0007_alter_animal_created_by_alter_animal_modified_by_and_more... OK
  Unapplying test_app.0006_team_admins_team_users... OK
  Unapplying test_app.0005_city... OK
  Unapplying test_app.0004_multiplefieldsmodel_animal... OK
  Unapplying test_app.0003_immutablelogentrynotcommon_immutablelogentry... OK
  Unapplying test_app.0002_set_up_resources_test_data... OK
  Unapplying dab_rbac.0001_initial... OK
2024-08-06 16:29:13,422 INFO     []  ansible_base.resource_registry.apps updating resource types
2024-08-06 16:29:13,422 INFO     []  ansible_base.resource_registry.apps updating resource types
2024-08-06 16:29:13,453 INFO     []  ansible_base.resource_registry.apps adding unmigrated resources for shared.user
2024-08-06 16:29:13,453 INFO     []  ansible_base.resource_registry.apps adding unmigrated resources for shared.user
2024-08-06 16:29:13,465 INFO     []  ansible_base.resource_registry.apps adding unmigrated resources for shared.team
2024-08-06 16:29:13,465 INFO     []  ansible_base.resource_registry.apps adding unmigrated resources for shared.team
2024-08-06 16:29:13,475 INFO     []  ansible_base.resource_registry.apps adding unmigrated resources for shared.organization
2024-08-06 16:29:13,475 INFO     []  ansible_base.resource_registry.apps adding unmigrated resources for shared.organization
2024-08-06 16:29:13,483 INFO     []  ansible_base.resource_registry.apps adding unmigrated resources for aap.authenticator
2024-08-06 16:29:13,483 INFO     []  ansible_base.resource_registry.apps adding unmigrated resources for aap.authenticator
Traceback (most recent call last):
  File "/home/alancoding/venvs/awx/lib64/python3.12/site-packages/django/apps/config.py", line 235, in get_model
    return self.models[model_name.lower()]
           ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^
KeyError: 'resourcemigrationtestmodel'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/alancoding/repos/awx/testing/django-ansible-base/manage.py", line 22, in <module>
    main()
  File "/home/alancoding/repos/awx/testing/django-ansible-base/manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "/home/alancoding/venvs/awx/lib64/python3.12/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/home/alancoding/venvs/awx/lib64/python3.12/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/alancoding/venvs/awx/lib64/python3.12/site-packages/django/core/management/base.py", line 412, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/alancoding/venvs/awx/lib64/python3.12/site-packages/django/core/management/base.py", line 458, in execute
    output = self.handle(*args, **options)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alancoding/venvs/awx/lib64/python3.12/site-packages/django/core/management/base.py", line 106, in wrapper
    res = handle_func(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alancoding/venvs/awx/lib64/python3.12/site-packages/django/core/management/commands/migrate.py", line 383, in handle
    emit_post_migrate_signal(
  File "/home/alancoding/venvs/awx/lib64/python3.12/site-packages/django/core/management/sql.py", line 52, in emit_post_migrate_signal
    models.signals.post_migrate.send(
  File "/home/alancoding/venvs/awx/lib64/python3.12/site-packages/django/dispatch/dispatcher.py", line 177, in send
    (receiver, receiver(signal=self, sender=sender, **named))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alancoding/repos/awx/testing/django-ansible-base/ansible_base/resource_registry/apps.py", line 72, in initialize_resources
    resource_model = apps.get_model(r_type.content_type.app_label, r_type.content_type.model)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alancoding/venvs/awx/lib64/python3.12/site-packages/django/apps/registry.py", line 213, in get_model
    return app_config.get_model(model_name, require_ready=require_ready)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alancoding/venvs/awx/lib64/python3.12/site-packages/django/apps/config.py", line 237, in get_model
    raise LookupError(
LookupError: App 'test_app' doesn't have a 'resourcemigrationtestmodel' model.
```

### PR DESCRIPTION
**Pull request by AlanCoding**
Fixes https://github.com/ansible/django-ansible-base/issues/555

### CODE REVIEW

**Comment by AlanCoding** [comment]
To throw out an alternative, depending on the feelings here, the error happens around here:

```python
        # Create resources
        for r_type in ResourceType.objects.all():
            resource_model = apps.get_model(r_type.content_type.app_label, r_type.content_type.model)
            resource_config = registry.get_config_for_model(resource_model)
```

So we are looking for a model that doesn't exist at the migration stage we're at, because that model has an entry in `ResourceType`. So why can't we just reverse the migration that created that `ResourceType` like normal? Because it wasn't created in a migration. It was created in a `post_migrate` hook, and herein lies a major flaw in attaching post-migrate logic. Honestly, none of our post-migrate methods will ever be valid for partial migrations.

Anyway, the alternative would be that we catch a `LookupError` that happens in the `apps.get_model` call. This signals to use that the `ResourceType` does not correspond to a real model. That could still be a real bug, so _then_ we call the method to see if we are fully-migrated or not, and re-`raise` only if we know we are fully migrated.

This alternative would be more performant, and more reactive as opposed to pro-active in this approach. But it's a fairly minor tweak to the overall philosophy.

File: ansible_base/resource_registry/apps.py [hunk 1]
```diff
@@ -10,6 +12,12 @@
 logger = logging.getLogger('ansible_base.resource_registry.apps')
 
 
+def migrations_are_complete():
+    executor = MigrationExecutor(connection)
+    plan = executor.migration_plan(executor.loader.graph.leaf_nodes())
+    return not bool(plan)
```
Comment (AlanCoding) [inline ansible_base/resource_registry/apps.py]: @relrod we would probably want to move this into utils? The same problems this is trying to fix likely exist somewhere in RBAC and maybe others.

Comment (relrod) [reply ansible_base/resource_registry/apps.py]: :+1: In #548 I introduce a lib.utils.db, that might be a good place? (if so, you can create that module here, I can deal with the conflict later)

File: ansible_base/lib/utils/db.py [hunk 2]
```diff
@@ -12,3 +13,15 @@ def ensure_transaction():
             yield
     else:
         yield
+
+
+def migrations_are_complete() -> bool:
+    """Returns a booling telling you if manage.py migrate has been run to completion
```
Comment (relrod) [inline ansible_base/lib/utils/db.py]:
```suggestion
    """Returns a boolean telling you if manage.py migrate has been run to completion
```

File: ansible_base/resource_registry/apps.py (hunk 1 above)
Comment (AlanCoding) [reply ansible_base/resource_registry/apps.py]: The rebase has merged these

**Comment by AlanCoding** [comment]
Documenting another side-effect error:

https://productionresultssa9.blob.core.windows.net/actions-results/d8bfb864-5d97-47b7-a7d6-16ef0c4b7f86/workflow-job-run-179484e7-c0ae-5a16-24ca-1a74f27caf55/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-08-23T01%3A29%3A31Z&sig=huyfmqfgTWEZSwmoRUNtTiqvvUz4LV83y%2BNN8O6WFJY%3D&ske=2024-08-23T12%3A23%3A13Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2024-08-23T00%3A23%3A13Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2024-05-04&sp=r&spr=https&sr=b&st=2024-08-23T01%3A19%3A26Z&sv=2024-05-04

```
2024-08-22T20:35:00.0384777Z You have 18 unapplied migration(s). Your project may not work properly until you apply the migrations for app(s): galaxy, sessions, social_django.
2024-08-22T20:35:00.0386391Z Run 'python manage.py migrate' to apply them.
2024-08-22T20:35:00.0387182Z     resource = Resource.get_resource_for_object(instance)
2024-08-22T20:35:00.0387942Z                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0389658Z   File "/usr/local/lib/python3.11/site-packages/ansible_base/resource_registry/models/resource.py", line 97, in get_resource_for_object
2024-08-22T20:35:00.0392396Z     return cls.objects.get(object_id=obj.pk, content_type=ContentType.objects.get_for_model(obj).pk)
2024-08-22T20:35:00.0393565Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0395035Z   File "/usr/local/lib/python3.11/site-packages/django/db/models/manager.py", line 87, in manager_method
2024-08-22T20:35:00.0396314Z     return getattr(self.get_queryset(), name)(*args, **kwargs)
2024-08-22T20:35:00.0397052Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0398151Z   File "/usr/local/lib/python3.11/site-packages/django/db/models/query.py", line 637, in get
2024-08-22T20:35:00.0399219Z     raise self.model.DoesNotExist(
2024-08-22T20:35:00.0400358Z ansible_base.resource_registry.models.resource.Resource.DoesNotExist: Resource matching query does not exist.
2024-08-22T20:35:00.0401685Z 
2024-08-22T20:35:00.0402096Z During handling of the above exception, another exception occurred:
2024-08-22T20:35:00.0402747Z 
2024-08-22T20:35:00.0402934Z Traceback (most recent call last):
2024-08-22T20:35:00.0403815Z   File "/usr/local/bin/pulpcore-manager", line 8, in <module>
2024-08-22T20:35:00.0404528Z     sys.exit(manage())
2024-08-22T20:35:00.0404948Z              ^^^^^^^^
2024-08-22T20:35:00.0405930Z   File "/usr/local/lib/python3.11/site-packages/pulpcore/app/manage.py", line 11, in manage
2024-08-22T20:35:00.0406943Z     execute_from_command_line(sys.argv)
2024-08-22T20:35:00.0408359Z   File "/usr/local/lib/python3.11/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
2024-08-22T20:35:00.0409618Z     utility.execute()
2024-08-22T20:35:00.0411090Z   File "/usr/local/lib/python3.11/site-packages/django/core/management/__init__.py", line 436, in execute
2024-08-22T20:35:00.0412396Z     self.fetch_command(subcommand).run_from_argv(self.argv)
2024-08-22T20:35:00.0413990Z   File "/usr/local/lib/python3.11/site-packages/django/core/management/base.py", line 412, in run_from_argv
2024-08-22T20:35:00.0415217Z     self.execute(*args, **cmd_options)
2024-08-22T20:35:00.0416880Z   File "/usr/local/lib/python3.11/site-packages/django/contrib/auth/management/commands/createsuperuser.py", line 88, in execute
2024-08-22T20:35:00.0418366Z     return super().execute(*args, **options)
2024-08-22T20:35:00.0419008Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0420304Z   File "/usr/local/lib/python3.11/site-packages/django/core/management/base.py", line 458, in execute
2024-08-22T20:35:00.0421480Z     output = self.handle(*args, **options)
2024-08-22T20:35:00.0422092Z              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0423682Z   File "/usr/local/lib/python3.11/site-packages/django/contrib/auth/management/commands/createsuperuser.py", line 233, in handle
2024-08-22T20:35:00.0425471Z     self.UserModel._default_manager.db_manager(database).create_superuser(
2024-08-22T20:35:00.0427173Z   File "/usr/local/lib/python3.11/site-packages/django/contrib/auth/models.py", line 172, in create_superuser
2024-08-22T20:35:00.0428567Z     return self._create_user(username, email, password, **extra_fields)
2024-08-22T20:35:00.0429497Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0430865Z   File "/usr/local/lib/python3.11/site-packages/django/contrib/auth/models.py", line 155, in _create_user
2024-08-22T20:35:00.0432376Z     user.save(using=self._db)
2024-08-22T20:35:00.0433570Z   File "/usr/local/lib/python3.11/site-packages/django/contrib/auth/base_user.py", line 76, in save
2024-08-22T20:35:00.0434613Z     super().save(*args, **kwargs)
2024-08-22T20:35:00.0435744Z   File "/usr/local/lib/python3.11/site-packages/django/db/models/base.py", line 814, in save
2024-08-22T20:35:00.0436764Z     self.save_base(
2024-08-22T20:35:00.0437903Z   File "/usr/local/lib/python3.11/site-packages/django/db/models/base.py", line 892, in save_base
2024-08-22T20:35:00.0438926Z     post_save.send(
2024-08-22T20:35:00.0440045Z   File "/usr/local/lib/python3.11/site-packages/django/dispatch/dispatcher.py", line 176, in send
2024-08-22T20:35:00.0441087Z     return [
2024-08-22T20:35:00.0441468Z            ^
2024-08-22T20:35:00.0442669Z   File "/usr/local/lib/python3.11/site-packages/django/dispatch/dispatcher.py", line 177, in <listcomp>
2024-08-22T20:35:00.0443960Z     (receiver, receiver(signal=self, sender=sender, **named))
2024-08-22T20:35:00.0444715Z                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0446248Z   File "/usr/local/lib/python3.11/site-packages/ansible_base/resource_registry/signals/handlers.py", line 34, in update_resource
2024-08-22T20:35:00.0447649Z     resource = init_resource_from_object(instance)
2024-08-22T20:35:00.0448412Z                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0450129Z   File "/usr/local/lib/python3.11/site-packages/ansible_base/resource_registry/models/resource.py", line 170, in init_resource_from_object
2024-08-22T20:35:00.0451910Z     resource_type = c_type.resource_type
2024-08-22T20:35:00.0452537Z                     ^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0454029Z   File "/usr/local/lib/python3.11/site-packages/django/db/models/fields/related_descriptors.py", line 492, in __get__
2024-08-22T20:35:00.0455415Z     raise self.RelatedObjectDoesNotExist(
2024-08-22T20:35:00.0456807Z django.contrib.contenttypes.models.ContentType.resource_type.RelatedObjectDoesNotExist: ContentType has no resource_type.
```

It looks like the error happens because resource_registry post-migrate logic runs _without ever having ran its own migrations_, which is a funny situation to be in.

**Review by john-westcott-iv — APPROVED** [review]

**Pull request merged**